Move variables and functions into new files

DucDung-oreo opened the pull request:
Add player.h
	Add player.cpp
	The game is having an issue that makes it crashed after pressing
	goButton, need to investigate more

DucDung-oreo committed:
Move variables and functions into new files

	Add player.h
	Add player.cpp
	The game is having an issue that makes it crashed after pressing
	goButton, need to investigate more

DucDung-oreo commented on mainwindow.cpp:
```diff
@@ -43,46 +43,28 @@ void MainWindow::on_scissorsButton_clicked()
+//    ui->rockButton->setEnabled(true);
+//    ui->scissorsButton->setEnabled(true);
+//    display_result();
+//    roundCount++;
```
comment out to debug the issue
after clicking the goButton, the program crashed.

DucDung-oreo replied:
me fixed that one.

DucDung-oreo committed:
Fix crashed bug in the program

	Add player1->initialize() to access the ui

ducbinh1996 commented on player.cpp:
```diff
@@ -0,0 +1,31 @@
+        ui->player1Label->setText("PAPER");
+        break;
+    case SCISSORS:
+        ui->player1Label->setText("SCISSORS");
```
These switch case is being used a lot in our code, I wonder if we can use unordered_map to reduce using switch case

DucDung-oreo replied:
We can use that. I'll update it in the next commit.

DucDung-oreo replied:
done

ducbinh1996 commented on player.h:
```diff
@@ -0,0 +1,22 @@
+    void intitialize(Ui::MainWindow *ui);
+    void display_choice(Choices player_choice);
+    Choices generate_choice();
+    Choices choice;
```
Make it private

DucDung-oreo replied:
ok me will do that

DucDung-oreo replied:
this'll mean we gotta use some things like a Vtable to access the private method, out side of the class, right?

ducbinh1996 replied:
On line 17, variable choice should be private, if you want to access it, use set and get method

DucDung-oreo replied:
OK got it

DucDung-oreo replied:
done

ducbinh1996 commented on player.cpp:
```diff
@@ -0,0 +1,31 @@
+    ui = ui_ptr;
+}
+
+void Player::display_choice(Choices player_choice)
```
I think this should be get_choice() function, the display thing should be handle in mainwindow (follow SOLID principle)

DucDung-oreo replied:
I'll change the name, and I'll check out that SOLID principle.

DucDung-oreo replied:
Moved display_round() and display_result() to mainwindow.

ducbinh1996 commented on mainwindow.cpp:
```diff
@@ -6,6 +6,7 @@ MainWindow::MainWindow(QWidget *parent)
     , ui(new Ui::MainWindow)
 {
     ui->setupUi(this);
+    player1->intitialize(ui);
```
why player2 not being initialize?

DucDung-oreo replied:
Me miss that
Me sorry
Me got it updated in the next commit

DucDung-oreo replied:
done

ducbinh1996 commented on mainwindow.cpp:
```diff
@@ -16,7 +17,7 @@ MainWindow::~MainWindow()
 {
-    player1Choice = ROCK;
+    player1->choice = ROCK;
     ui->paperButton->setEnabled(false);
```
what if I want to change my choice? why we not let the button always enable?

DucDung-oreo replied:
gud idea
imma disable the goButton until we click one of the three choice buttons
and the choice buttons are not disable

DucDung-oreo replied:
done

ducbinh1996 commented on mainwindow.h:
```diff
@@ -26,16 +25,15 @@ private slots:
     void display_round();
     void display_result();
+
+    Player* player1 = new Player;
```
it's okay for now, but if we support multiple players then I think we should change this

DucDung-oreo replied:
ok
but let's just keep it that way for now.

DucDung-oreo replied:
Gonna mark this one as done.

ducbinh1996 commented on player.cpp:
```diff
@@ -0,0 +1,31 @@
+
+void Player::intitialize(Ui::MainWindow *ui_ptr)
+{
+    ui = ui_ptr;
```
if we add another ui into our game then this will be a problem (if we forget to re-initialize)

DucDung-oreo replied:
okay, me gotta tell meself that me understand that

ducbinh1996 commented on player.h:
```diff
@@ -0,0 +1,22 @@
+#include "ui_mainwindow.h"
+#include <QMainWindow>
+
+enum Choices { ROCK, PAPER, SCISSORS };
```
If player does not choose anything before click go button, then it always being rock (not a good design)

DucDung-oreo replied:
Thanks, so unordered_map is gonna fix that, right

ducbinh1996 replied:
Best practices that we can add invalid_choice=0 to Choices enum, then we can remind user to choose before click go button

DucDung-oreo replied:
ok, gud lời khuyên

ducbinh1996 commented on mainwindow.cpp:
```diff
@@ -56,35 +59,15 @@ void MainWindow::on_goButton_clicked()
-
 QString MainWindow::get_result(Choices p1, Choices p2)
 {
     if (p1 == p2)
```
why we need to calculate winCount?

DucDung-oreo replied:
what do you mean?
I want to have winCount, loseCount and tieCount, but roundCount =  winCount + loseCount + tieCount
so instead of increasing winCount, I just calculate it through tieCount  and loseCount

ducbinh1996 replied:
I mean it is difficult to track the code flow when you doing like this, my suggest is to move this logic to where you handle tieCount and loseCount

DucDung-oreo replied:
hmm ok, imma move it somewhere else

DucDung-oreo replied:
Done.

ducbinh1996 commented on mainwindow.h:
```diff
@@ -26,16 +25,15 @@ private slots:
-    Choices player1Choice;
-    Choices player2Choice;
     int roundCount = 1;
     int winCount = 0;
```
I believe these values are belong to player class :))

DucDung-oreo replied:
I have moved them into player class.

DucDung-oreo replied:
Done

ducbinh1996 reviewed (changes requested):
Please check my comments

DucDung-oreo committed:
Fix some issue in the PR - part 1

	Use unordered_map instead of enum
	Update display_choice() with QLabel parameter
	Add initialize() for player2
	Add done_chosing() function
	Update disable/enable rule

ducbinh1996 commented on player.cpp:
```diff
@@ -8,24 +8,26 @@ Player::Player()
 void Player::intitialize(Ui::MainWindow *ui_ptr)
 {
     ui = ui_ptr;
+    choices_map =
```
I this this map should be a const

DucDung-oreo replied:
ok noted

DucDung-oreo replied:
cannot make it work right row
me thinks we should fix it later

DucDung-oreo replied:
an issue have been created for this comment: https://github.com/DucDung-oreo/rock_paper_scissors/issues/4

ducbinh1996 commented on mainwindow.cpp:
```diff
@@ -17,58 +19,44 @@ MainWindow::~MainWindow()
-    player1->choice = PAPER;
-    ui->rockButton->setEnabled(false);
-    ui->scissorsButton->setEnabled(false);
+    player1->choice = 1;
```
I think we can keep the choice enum and use it in the map instead of using number

DucDung-oreo replied:
got it

DucDung-oreo replied:
done.

DucDung-oreo committed:
Fix for PR comment - part 2

	Move choice variable into private
	Add set_choice() and get_choice()
	Keep using enum Choices intead of int

DucDung-oreo committed:
Fix for PR comment - part 3

	Add destructor for player1 and player 2
	Keep using enum as key in unordered_map
	Remove unwanted line
	Fix spelling error

DucDung-oreo committed:
Fix for PR's comment - part 4

	Move winCount, tieCount and loseCount to class Player
	Add getter for them

	ps: I think this would help future use when comparing the
	winCount of each player to determine the winner.

DucDung-oreo committed:
Fix for PR's comment - part 5

	Merge display_round() with display_result()
	Use QTimer and connect() to display round and result every 500ms
